### PR TITLE
Add event listeners via service loader mechanism

### PR DESCRIPTION
**Pull request by felixbarny**
Also improve extendability of LoggingEventListener.

### CODE REVIEW

**Comment by felixbarny** [comment]
To continue on our discussion on the dev mailing list, I think that people wanting to customize the logger should extend `LoggingEventListener` rather than implement `P6Logger`. For example, if you want to only log if a wild exception appears, just override logElapsed:

```java
public class ErrorLoggingEventListener extends LoggingEventListener {
  @Override
  protected void logElapsed(Loggable loggable, long timeElapsedNanos, Category category, SQLException e) {
    if (e != null) {
      super.logElapsed(loggable, timeElapsedNanos, category, e);
    }
  }
}
```

And create `src/main/resources/META-INF/services/com.p6spy.engine.event.JdbcEventListener`:
```
org.example.ErrorLoggingEventListener
```

**Comment by felixbarny** [comment]
Still a bit unsure how to exclude the original LoggingEventListener then...

**Comment by felixbarny** [comment]
thoughts?

**Comment by quintonm** [comment]
The LoggingEventListener is enabled through P6LogFactory, correct?  If so, then to disable LoggingEventListener, one simply needs to exclude P6LogFactory from modulelist.

**Comment by quintonm** [comment]
The service registration changes look good to me... Just need a test to show that it works as expected.

**Comment by felixbarny** [comment]
IIRC, excluding the logging module also causes the options to not be initialized. Could't this lead to problems?

**Comment by quintonm** [comment]
Yes, I did not think about that part.  I suppose that if someone wants to alter logging, they must extend the logger or replace the logging module.  However, I think that is a separate issue to the service loading mechanism.

**Comment by felixbarny** [comment]
It is indeed. Pretty much going on at work right now so the test will probably have to wait until the weekend.

**Comment by felixbarny** [comment]
Added the test and found a solution to add a custom LoggingEventListener via another service loader.

File: src/main/java/com/p6spy/engine/logging/ErrorLoggingEventListener.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+package com.p6spy.engine.logging;
+
+import com.p6spy.engine.common.Loggable;
+
+import java.sql.SQLException;
+
+/**
+ * Extends {@link LoggingEventListener} and only logs if an exception has occurred.
+ * <p/>
+ * To activate this class, add <code>com.p6spy.engine.logging.LoggingEventListener</code> to the file
```
Comment (quintonm) [inline src/main/java/com/p6spy/engine/logging/ErrorLoggingEventListener.java]: Shouldn't this be com.p6spy.engine.logging.ErrorLoggingEventListener?

**Comment by quintonm** [comment]
Except for the one possible problem with the javadocs, it looks really good.

**Comment by quintonm** [comment]
solves #332

File: src/main/java/com/p6spy/engine/logging/ErrorLoggingEventListener.java (hunk 1 above)
Comment (felixbarny) [inline src/main/java/com/p6spy/engine/logging/ErrorLoggingEventListener.java]: You're right!

**Comment by felixbarny** [comment]
I don't quite get what's wrong with the tests:

```
$ script/travis/before_script_firebird.sh
ISQL Version: LI-V2.5.1.26351 Firebird 2.5
Use CONNECT or CREATE DATABASE to specify a database
Statement failed, SQLSTATE = 08006
Unable to complete network request to host "localhost".
-Failed to locate host machine.
After line 0 in file /tmp/firebird_create_db.sql
```

**Comment by typekpb** [comment]
well, not sure, but I'd give it a try with changing a contents of: `/script/travis/before_script_firebird.sh`
from:
`echo "create database '/tmp/firebird_p6spy.gdb';" > /tmp/firebird_create_db.sql # preps for test DB creation`
to:
`echo "create database '127.0.0.1:/tmp/firebird_p6spy.gdb';" > /tmp/firebird_create_db.sql # preps for test DB creation`
If there is a problem with name resolution (localhost).

honestly: I didn't give it a try and be ready to roll it back once it has no/bad impact.

**Comment by felixbarny** [comment]
I think we should do that in another pull request. The current master build is also failing because of this error so it does not have to do with my changes. So I think we can merge this PR

**Comment by quintonm** [comment]
Agreed.

**Pull request merged**